File: client/src/app/services/game-play/replay.service.ts
import { EventEmitter, Injectable, OnDestroy } from '@angular/core';
import { NB_MS_IN_SECOND, TICK } from '@app/constants/time-constants';
import { HttpService } from '@app/services/communication/http.service';
import { DelayService } from '@app/services/divers/delay.service';
import { GameDataService } from '@app/services/game-play/game-data.service';
import { SoundService } from '@app/services/game-play/sound.service';
import { PrivateUserDataDTO } from '@common/dto/user/private-user-data.dto';
import { GameMode } from '@common/enums/game-play/game-mode';
import { Speed } from '@common/enums/game-play/speed';
import {
    GameInformationForReplay,
    GameStateSnapshot,
    Replay,
    ReplayAction,
    ReplayInformationToRestore,
    ReplayInput,
} from '@common/interfaces/game-play/replay-action';
import { GameTimeService } from './game-time.service';
import { GameService } from './game.service';

@Injectable({
    providedIn: 'root',
})
export class ReplayService implements OnDestroy {
    isReplayMode: boolean;
    isReplaying: boolean;
    storeActions: boolean;
    replayEvent: EventEmitter<void>;
    replayActionTrigger: EventEmitter<{ category: string; input: ReplayInput }> = new EventEmitter();

    private gameService: GameService;
    private snapshots: GameStateSnapshot[] = [];
    private finalTime: number = 0;
    private replayActions: ReplayAction[];
    private actionIndex: number;
    private gameDetails: GameInformationForReplay;
    private replayInformationToRestore: ReplayInformationToRestore;
    private userData: PrivateUserDataDTO;

    // eslint-disable-next-line max-params
    constructor(
        private delayService: DelayService,
        private soundService: SoundService,
        private gameTimeService: GameTimeService,
        private gameDataService: GameDataService,
        private httpService: HttpService,
    ) {
        this.replayEvent = new EventEmitter<void>();
        this.reset();
    }

    get getterReplayInformationToRestore(): ReplayInformationToRestore {
        return this.replayInformationToRestore;
    }

    init(userData: PrivateUserDataDTO): void {
        this.userData = userData;
    }

    resetEverything(): void {
        this.snapshots = [];
        this.replayActions = [];
        this.finalTime = 0;
        this.actionIndex = 0;
        this.gameDetails = {
            recordedTimes: [],
            players: [],
            totalTime: 0,
            initialTime: 0,
            elapsedTime: 0,
            originalUrl: '',
            modifiedUrl: '',
            finalTime: 0,
            gameName: '',
            difficulty: '',
            gameId: '',
            nbOfPlayers: 0,
            gameMode: GameMode.None,
            totalDifferences: 0,
            gameValues: {
                timerTime: 0,
                penaltyTime: 0,
                gainedTime: 0,
            },
            timeToStart: 0,
            originalDifferencesUrls: [],
            flashingDifferencesUrls: [],
            cheatImages: [],
            isObserving: false,
        };
        this.replayInformationToRestore = {
            gameDetails: this.gameDetails,
            actions: [],
            snapshots: [],
        };
        this.isReplayMode = false;
        this.isReplaying = false;
        this.storeActions = true;
        this.gameTimeService.isReplayMode = false;
        this.gameTimeService.recordedTimeIndex = 0;
        this.gameTimeService.recordedTimes = [];
        this.gameDataService.isObserving = false;
        this.gameDataService.timeToStart = 0;
        this.gameDataService.playerNames = [];
        this.gameDataService.chronoTime = 0;
        this.gameDataService.originalPicture = '';
        this.gameDataService.modifiedPicture = '';
        this.gameDataService.gameID = '';
        this.gameDataService.nbOfPlayers = 0;
        this.gameDataService.differenceNbr = 0;
        this.gameDataService.difficulty = '';
        this.gameDataService.gameName = '';
        this.gameDataService.gameMode = GameMode.None;
        this.gameDataService.gameValues = {
            timerTime: 0,
            penaltyTime: 0,
            gainedTime: 0,
        };
        this.gameDataService.playerScores = [];
        this.gameDataService.foundDifferences = [];
        this.gameDataService.setOriginalDifferencesUrls([]);
        this.gameDataService.setFlashingDifferencesUrls([]);
        this.gameService?.cleanup();
    }

    setReplay(replay: Replay): void {
        this.snapshots = replay.snapshots;
        this.replayActions = replay.actions;
        this.gameDetails = replay.gameDetails;
        this.finalTime = this.gameDetails.finalTime;
        this.replayInformationToRestore = {
            gameDetails: this.gameDetails,
            actions: replay.actions,
            snapshots: replay.snapshots,
        };
        this.isReplayMode = true;
        this.setGameDetailsThroughoutAllTheGame(this.gameDetails);
        this.replay();
    }

    setGameDetailsThroughoutAllTheGame(gameDetails: GameInformationForReplay): void {
        this.gameDataService.gameID = gameDetails.gameId;
        this.gameDataService.playerNames = gameDetails.players;
        this.gameDataService.chronoTime = gameDetails.totalTime;
        this.gameDataService.originalPicture = gameDetails.originalUrl;
        this.gameDataService.modifiedPicture = gameDetails.modifiedUrl;
        this.gameDataService.gameName = gameDetails.gameName;
        this.gameDataService.difficulty = gameDetails.difficulty;
        this.gameDataService.nbOfPlayers = gameDetails.nbOfPlayers;
        this.gameDataService.gameMode = gameDetails.gameMode;
        this.gameDataService.differenceNbr = gameDetails.totalDifferences;
        this.gameDataService.gameValues = gameDetails.gameValues;
        this.gameDataService.timeToStart = gameDetails.timeToStart;
        this.gameDataService.setOriginalDifferencesUrls(gameDetails.originalDifferencesUrls);
        this.gameDataService.setFlashingDifferencesUrls(gameDetails.flashingDifferencesUrls);
        this.gameService.totalDifferences = gameDetails.totalDifferences;
    }

    initializeGameDetails(): void {
        this.gameDetails = {
            recordedTimes: this.gameTimeService.recordedTimes,
            players: this.gameDataService.playerNames,
            totalTime: this.gameDataService.chronoTime,
            initialTime: this.gameTimeService.getInitialTime(),
            elapsedTime: this.getElapsedTime(),
            originalUrl: this.gameDataService.originalPicture,
            modifiedUrl: this.gameDataService.modifiedPicture,
            finalTime: this.finalTime,
            gameName: this.gameDataService.gameName,
            difficulty: this.gameDataService.difficulty,
            gameId: this.gameDataService.gameID,
            nbOfPlayers: this.gameDataService.nbOfPlayers,
            gameMode: this.gameDataService.gameMode,
            totalDifferences: this.gameDataService.differenceNbr,
            gameValues: this.gameDataService.gameValues,
            timeToStart: this.gameDataService.timeToStart,
            originalDifferencesUrls: this.gameDataService.getOriginalDifferencesUrls(),
            flashingDifferencesUrls: this.gameDataService.getFlashingDifferencesUrls(),
            cheatImages: [],
            isObserving: this.gameDataService.isObserving,
        };
    }

    ngOnDestroy(): void {
        this.gameService?.cleanup();
    }

    setGameService(gameService: GameService) {
        if (!this.gameService) {
            this.gameService = gameService;
        }
    }

    getInitialTime(): number {
        return this.gameTimeService.getInitialTime();
    }

    getElapsedTime(): number {
        return this.gameTimeService.getInitialTime() - this.finalTime;
    }

    getCurrentGameState(): GameStateSnapshot {
        return {
            time: this.gameTimeService.replayTime,
            state: {
                totalDifferences: this.gameService?.totalDifferences,
                differencesFoundTotal: this.gameService?.differencesFoundTotal,
                originalDifferencesUrls: this.gameDataService.getOriginalDifferencesUrls(),
            },
        };
    }

    captureSnapshot(): void {
        if (this.isReplayMode) {
            return;
        }
        const currentState = this.getCurrentGameState();
        this.snapshots.push(currentState);
    }

    restoreGameStateFromTime(time: number): void {
        const snapshotIndex = this.findNextSnapshot(time);
        if (snapshotIndex < 0 || snapshotIndex >= this.snapshots.length) {
            return;
        }

        const snapshot = this.snapshots[snapshotIndex];
        this.restoreGameState({ time: snapshot.time, state: snapshot.state });
        this.resetActionIndexDependingOnTime(time);
    }

    restoreGameState(snapshotState: GameStateSnapshot): void {
        if (this.gameService) {
            this.gameService.totalDifferences = snapshotState.state.totalDifferences;
            this.gameService.differencesFoundTotal = snapshotState.state.differencesFoundTotal;
            this.gameDataService.setOriginalDifferencesUrls(snapshotState.state.originalDifferencesUrls);
        }
    }

    findNextSnapshot(time: number): number {
        let index = -1;
        let currentHighestTime = 0;
        for (let i = 0; i < this.snapshots.length; i++) {
            if (typeof this.snapshots[i].time === 'undefined') {
                continue;
            }
            if (this.snapshots[i].time <= time && this.snapshots[i].time > currentHighestTime) {
                index = i;
                currentHighestTime = this.snapshots[i].time;
                break;
            }
        }
        return index;
    }

    resetActionIndexDependingOnTime(time: number): void {
        this.actionIndex = this.findIndexBeforeNextAction(time);
    }

    findIndexBeforeNextAction(time: number): number {
        let index = 0;
        while (index < this.replayActions.length && this.replayActions[index].time >= time) {
            index++;
        }
        return index;
    }

    reset(): void {
        this.stop();
        this.delayService.timeIsPaused = false;
        this.delayService.changeSpeed(Speed.NORMAL);
        this.gameTimeService.recordedTimes = [];
        this.replayActions = [];
        this.isReplaying = false;
        this.storeActions = true;
        this.isReplayMode = false;
    }

    doAndStore(category: string, input: ReplayInput) {
        if (!this.isReplayMode) {
            this.captureSnapshot();
            this.store(category, input);
            this.doAction(category, input);
            this.captureSnapshot();
        }
    }

    doAction(category: string, input: ReplayInput) {
        this.replayActionTrigger.emit({ category, input });
    }

    store(category: string, input: ReplayInput) {
        if (this.storeActions && !this.isReplayMode) {
            this.captureSnapshot();
            this.replayActions.push({ time: this.gameTimeService.replayTime, category, input });
            this.finalTime = this.gameTimeService.replayTime;
        }
    }

    bundleReplayForPostRequest(): Replay {
        const replay: Replay = {
            gameDetails: this.gameDetails,
            actions: this.replayActions,
            snapshots: this.snapshots,
            createdAt: new Date(),
            createdBy: this.userData.username,
        };
        return replay;
    }

    postReplay(): void {
        this.initializeGameDetails();
        this.httpService.postReplay(this.bundleReplayForPostRequest());
    }

    pause(): void {
        this.delayService.pauseTime();
        this.soundService.pause();
    }

    resume(): void {
        this.delayService.resumeTime();
        this.soundService.resume();
    }

    restart(): void {
        this.stop();
        this.replay();
    }

    stop() {
        this.endOfReplay();
        this.soundService.pause();
        this.delayService.clearDelays();
    }

    endOfReplay() {
        this.isReplaying = false;
        this.delayService.clearCycles();
    }

    changeSpeed(speed: Speed) {
        this.delayService.changeSpeed(speed);
        this.soundService.speed = speed;
    }

    private tick() {
        this.gameTimeService.nextRecordedTime();
        let replayAction: ReplayAction;
        do {
            replayAction = this.replayActions[this.actionIndex];
            if (this.isSameTime(replayAction.time, this.gameTimeService.replayTime)) {
                this.actionIndex++;
                this.doAction(replayAction.category, replayAction.input);
            }
        } while (this.actionIndex < this.replayActions.length && this.isSameTime(replayAction.time, this.gameTimeService.replayTime));
    }

    private async replay(): Promise<void> {
        this.delayService.timeIsPaused = false;
        this.gameTimeService.isReplayMode = true;
        this.gameTimeService.recordedTimeIndex = 0;
        this.replayEvent.emit();
        this.isReplayMode = true;
        this.isReplaying = true;
        this.storeActions = false;
        this.actionIndex = 0;
        this.delayService.doCyclically(TICK * NB_MS_IN_SECOND, () => this.tick());
    }

    private isSameTime(time1: number, time2: number): boolean {
        const precision = 1e-3;
        return Math.abs(time1 - time2) < precision;
    }
}

File: flutter_project/lib/service/replay-service.dart
// ignore_for_file: unused_label, unnecessary_import
import 'dart:async';
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_project/classes/gamePlay/differences-images.dart';
import 'package:flutter_project/classes/gamePlay/game-values.dart';
import 'package:flutter_project/classes/gamePlay/player-info.dart';
import 'package:flutter_project/common/enum.dart';
import 'package:flutter_project/common/replay-constants.dart';
import 'package:flutter_project/dtos/replay-dto.dart';
import 'package:flutter_project/environnment.dart';
import 'package:flutter_project/service/authentication-service.dart';
import 'package:flutter_project/service/delay-service.dart';
import 'package:flutter_project/service/game-data-service.dart';
import 'package:flutter_project/service/game-service.dart';
import 'package:flutter_project/service/sound-service.dart';
import 'package:http/http.dart' as http;

class ReplayService with ChangeNotifier {
  final String baseUrl = Environnment.httpLink;

  DelayService _delayService = DelayService();
  GameService _gameService = GameService();
  SoundService _soundService = SoundService();

  bool isReplayMode = false;
  bool isReplayingReplay = false;
  bool isReplaying = false;
  bool storeActions = true;

  List<GameStateSnapshot> snapshots = [];
  double finalTime = 0;
  get getFinalTime => _gameService.getFinalRecordedTime();
  List<ReplayAction> replayActions = [];
  int actionIndex = 0;
  GameInformationForReplay gameDetails = GameInformationForReplay(
    recordedTimes: [],
    players: [],
    totalTime: 0.0,
    initialTime: 0.0,
    elapsedTime: 0.0,
    originalUrl: '',
    modifiedUrl: '',
    finalTime: 0.0,
    gameName: '',
    difficulty: '',
    gameId: '',
    nbOfPlayers: 0,
    gameMode: GameMode.ClassicDeathMatch,
    totalDifferences: 0,
    gameValues: GameValues(),
    timeToStart: 0,
    originalDifferencesUrls: [],
    flashingDifferencesUrls: [],
    cheatImages: [],
    isObserving: false,
  );

  ReplayInformationToRestore replayInformationToRestore =
      ReplayInformationToRestore(
          gameDetails: GameInformationForReplay(
            recordedTimes: [],
            players: [],
            totalTime: 0.0,
            initialTime: 0.0,
            elapsedTime: 0.0,
            originalUrl: '',
            modifiedUrl: '',
            finalTime: 0.0,
            gameName: '',
            difficulty: '',
            gameId: '',
            nbOfPlayers: 0,
            gameMode: GameMode.ClassicDeathMatch,
            totalDifferences: 0,
            gameValues: GameValues(),
            timeToStart: 0,
            originalDifferencesUrls: [],
            flashingDifferencesUrls: [],
            cheatImages: [],
            isObserving: false,
          ),
          actions: [],
          snapshots: []);

  StreamController<void> replayEvent = StreamController.broadcast();
  StreamController<ReplayAction> replayActionTrigger =
      StreamController<ReplayAction>.broadcast();
  StreamController<double> _replayTimeController =
      StreamController<double>.broadcast();
  StreamController<bool> isResetReplayController =
      StreamController<bool>.broadcast();

  Stream<double> get replayTimeStream => _replayTimeController.stream;
  Stream<ReplayAction> get replayActionTriggered => replayActionTrigger.stream;
  Stream<bool> get isResetReplayStream => isResetReplayController.stream;

  static final ReplayService _instance = ReplayService._internal(
      DelayService(), GameService.instance, SoundService());

  factory ReplayService() {
    return _instance;
  }

  static ReplayService get instance => _instance;

  ReplayService._internal(
      this._delayService, this._gameService, this._soundService);

  void closeStreams() {
    replayEvent.close();
    replayActionTrigger.close();
    _replayTimeController.close();
  }

  void reinitializeStreams() {
    closeStreams();
    replayEvent = StreamController.broadcast();
    replayActionTrigger = StreamController<ReplayAction>.broadcast();
    _replayTimeController = StreamController<double>.broadcast();
  }

  void initialize(DelayService delayService, GameService gameService,
      SoundService soundService, GameDataService gameDataService) {
    _delayService = delayService;
    _gameService = gameService;
    _soundService = soundService;
  }

  void resetEveryValueInTheGame() {
    gameDetails.elapsedTime = 0;
    gameDetails.totalTime = 0;
    gameDetails.finalTime = 0;
    gameDetails.initialTime = 0;
    _gameService.recordedTimes = [];

    _gameService.differencesFoundTotal = 0;
    _gameService.totalDifferences = 0;
    _gameService.time = getInitialTime();
    _gameService.recordedTimeIndex = 0;
    snapshots = [];
    replayActions = [];
    actionIndex = 0;
    _gameService.gameData.playersInfo.forEach((player) {
      if (player.username != "Observer") {
        player.score = 0;
      }
    });
  }

  double getCurrentTime() {
    return _gameService.time;
  }

  void setDifferenceOverlay(DifferenceImages newValue) {
    _gameService.updateDifferenceOverlay(newValue);
    notifyListeners();
  }

  void setReplay(Replay replay) {
    gameDetails = replay.gameDetails;
    replayActions = replay.actions;
    snapshots = replay.snapshots;
    finalTime = replay.gameDetails.finalTime;
    replayInformationToRestore = ReplayInformationToRestore(
        gameDetails: replay.gameDetails,
        actions: replay.actions,
        snapshots: replay.snapshots);
    isReplayMode = true;
    //print("running setReplayThroughAllTheGame");

    setGameDetailsThroughoutAllTheGame(gameDetails);

    restart();
  }

  int getDifficulty(String difficulty) {
    if (difficulty == 'facile') {
      return 0;
    } else if (difficulty == 'difficile') {
      return 1;
    } else {
      return 0;
    }
  }

  void setGameDetailsThroughoutAllTheGame(
      GameInformationForReplay gameDetails) {
    GameDataService.instance.gameID = gameDetails.gameId;
    GameDataService.instance.playersInfo = gameDetails.players
        .map((playerName) => PlayerInfo(username: playerName))
        .toList();
    GameDataService.instance.chronometerTime = gameDetails.totalTime;
    GameDataService.instance.originalPicture = gameDetails.originalUrl;
    replayActionTrigger.add(ReplayAction(
        time: getInitialTime(),
        category: 'startOriginalPicture',
        input: ReplayInput<String>(value: gameDetails.originalUrl)));

    GameService.instance.originalImageLimited.add(gameDetails.originalUrl);

    GameService.instance.modifiedImageLimited.add(gameDetails.modifiedUrl);

    GameDataService.instance.modifiedPicture = gameDetails.modifiedUrl;
    replayActionTrigger.add(ReplayAction(
        time: getInitialTime(),
        category: 'startModifiedPicture',
        input: ReplayInput<String>(value: gameDetails.modifiedUrl)));
    GameDataService.instance.gameName = gameDetails.gameName;
    GameDataService.instance.difficulty = getDifficulty(gameDetails.difficulty);
    GameDataService.instance.nbOfPlayers = gameDetails.nbOfPlayers;
    GameDataService.instance.gameMode = gameDetails.gameMode;
    GameDataService.instance.differenceNbr = gameDetails.totalDifferences;
    GameDataService.instance.gameValues = gameDetails.gameValues;
    GameDataService.instance.timeToStart = gameDetails.timeToStart;
    GameService.instance.latestDifferenceOverlay =
        gameDetails.originalDifferencesUrls;
    GameService.instance.latestFlickerEvent =
        gameDetails.flashingDifferencesUrls;
    GameService.instance.totalDifferences = gameDetails.totalDifferences;
    GameService.instance.differencesFoundTotal = 0;
    GameDataService.instance.isObserver = gameDetails.isObserving;
    GameService.instance.recordedTimes = gameDetails.recordedTimes;
    GameService.instance.recordedTimeIndex = 0;
  }

  void initializeGameDetails() {
    gameDetails = GameInformationForReplay(
      recordedTimes: _gameService.recordedTimes,
      players: GameDataService.instance.playersInfo
          .map((player) => player.username)
          .toList(),
      totalTime: getInitialTime(),
      initialTime: getInitialTime(),
      elapsedTime: getElapsedTime(),
      originalUrl: GameService.instance.originalImageLimited.first,
      modifiedUrl: GameService.instance.modifiedImageLimited.first,
      finalTime: finalTime,
      gameName: GameDataService.instance.gameName,
      difficulty: GameDataService.instance.difficulty.toString(),
      gameId: GameDataService.instance.gameID,
      nbOfPlayers: GameDataService.instance.nbOfPlayers,
      gameMode: GameDataService.instance.gameMode,
      totalDifferences: GameDataService.instance.differenceNbr,
      gameValues: GameDataService.instance.gameValues,
      timeToStart: GameDataService.instance.timeToStart,
      originalDifferencesUrls: GameService.instance.latestDifferenceOverlay,
      flashingDifferencesUrls: GameService.instance.latestFlickerEvent,
      cheatImages: GameService.instance.latestFlickerEvent,
      isObserving: GameDataService.instance.isObserver,
    );
    print("gameDetails");
    print(gameDetails.originalUrl);
  }

  double getInitialTime() {
    return _gameService.getInitialTime();
  }

  double getElapsedTime() {
    return getInitialTime() - getFinalTime;
  }

  GameStateSnapshot getCurrentGameState() {
    return GameStateSnapshot(
      time: _gameService.time,
      state: GameState(
        totalDifferences: GameService.instance.totalDifferences,
        differencesFoundTotal: GameService.instance.differencesFoundTotal,
        originalDifferencesUrls: GameService.instance.latestDifferenceOverlay,
      ),
    );
  }

  void captureSnapshot() {
    if (isReplayMode) {
      return;
    }
    snapshots.add(getCurrentGameState());
  }

  void restoreGameStateFromTime(double time) {
    //print("in restoreGameStateFromTime");
    //print("time: $time");

    int snapshotIndex = findNextSnapshot(time);
    if (snapshotIndex < 0 || snapshotIndex >= snapshots.length) {
      return;
    } else {
      restoreGameState(snapshots[snapshotIndex]);
      GameService.instance.differenceOverlayController.close();
      //undoLastSnapShot();
      //restoreSnapshotBefore(time);

      GameService.instance.differenceOverlayController =
          StreamController<String>.broadcast();
      isResetReplayController.add(true);
      restoreGameState(snapshots[snapshotIndex]);
      resetActionIndexDependingOnTime(time);
    }
  }

  void restoreSnapshotBefore(double time) {
    int index = 0;
    double currentClosestTime = 0.0;
    for (int i = 0; i < snapshots.length; i++) {
      if (snapshots[i].time >= time && snapshots[i].time < currentClosestTime) {
        print("WENT IN");
        index = i;
        currentClosestTime = snapshots[i].time;
      }
    }

    List<String> differencesUrls =
        snapshots[index].state.originalDifferencesUrls;

    print("RESTORESNAPSHOTBEFORE");

    for (int i = 0; i < differencesUrls.length; i++) {
      if (differencesUrls != []) {
        print("RestoreSnapShot");
        GameService.instance.differenceOverlayController.sink
            .add(differencesUrls[i]);
      }
    }
  }

  void undoLastSnapShot() {
    List<String> differencesUrls = snapshots.last.state.originalDifferencesUrls;

    for (int i = 0; i < differencesUrls.length; i++) {
      print("UNDOLASTSNAPSHOT");
      if (differencesUrls.isNotEmpty) {
        GameService.instance.differenceOverlayController.sink
            .add(differencesUrls[i]);
      }
    }
  }

  void restoreGameState(GameStateSnapshot snapshot) {
    _gameService.totalDifferences = snapshot.state.totalDifferences;
    _gameService.differencesFoundTotal = snapshot.state.differencesFoundTotal;
  }

  int findNextSnapshot(double time) {
    int index = -1;
    double currentHighestTime = 0.0;
    for (int i = 0; i < snapshots.length; i++) {
      if (snapshots[i].time <= time && snapshots[i].time > currentHighestTime) {
        index = i;
        currentHighestTime = snapshots[i].time;
      }
    }
    return index;
  }

  resetActionIndexDependingOnTime(double Time) {
    actionIndex = findIndexBeforeNextAction(Time);
  }

  int findIndexBeforeNextAction(double time) {
    int index = 0;
    while (index < replayActions.length && replayActions[index].time >= time) {
      index++;
    }
    return index;
  }

  void reset() {
    stop();
    _delayService.timeIsPaused = false;
    _delayService.changeSpeed(Speed.NORMAL);
    _gameService.recordedTimes = [];
    replayActions = [];
    isReplaying = false;
    storeActions = true;
    isReplayMode = false;
  }

  void doAction(String category, dynamic input) {
    //print("in do action");
    replayActionTrigger.add(ReplayAction(
        time: _gameService.time, category: category, input: input));
  }

  void store(String category, ReplayInput<dynamic> input) {
    //print("in store");
    //print("isReplaying: $isReplaying");
    //print("storeActions: $storeActions");
    if (storeActions && !isReplaying) {
      captureSnapshot();

      replayActions.add(ReplayAction(
          time: _gameService.recordedTimes.last,
          category: category,
          input: input));
      finalTime = _gameService.time;
    }
  }

  doAndStore(String category, ReplayInput input) {
    isResetReplayController.add(false);
    if (!isReplayMode) {
      captureSnapshot();
      store(category, input);
      doAction(category, input);
      captureSnapshot();
    }
  }

  Replay bundleReplayForPostRequest() {
    return Replay(
        gameDetails: gameDetails,
        actions: replayActions,
        snapshots: snapshots,
        createdAt: DateTime.now(),
        createdBy: AuthService.instance.username);
  }

  Future<Replay> requestToPostReplay() async {
    Replay replay = bundleReplayForPostRequest();
    final token = AuthService.instance.token;
    print("tttttttttttttttt");
    print(replay.toJson());
    var json = replay.toJson();
    print(jsonEncode(json));

    final response = await http.post(
      Uri.parse('$baseUrl/replays'),
      body: jsonEncode({"replay": replay.toJson()}),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
    );
    print(response);

    if (response.statusCode == 200) {
      print("reussi");
      return replay;
    } else {
      throw Exception('Failed to post replay: ${response.body}');
    }
  }

  void postReplay() {
    initializeGameDetails();
    requestToPostReplay();
  }

  void pause() {
    _delayService.pauseTime();
    _soundService.pause();
    print("PAUSE");
    GameService().cheatDataController.add([]);
    GameDataService().showCheat = false;
  }

  void resume() {
    _delayService.resumeTime();
    _soundService.resume();
  }

  void restart() {
    _gameService.gameData.playersInfo.forEach((player) {
      player.score = 0;
    });
    //restoreGameStateFromTime(getInitialTime());
    //print("in restart");
    stop();
    doReplay();
  }

  void stop() {
    endOfReplay();
    _soundService.pause();
    _delayService.clearDelays();
  }

  endOfReplay() {
    isReplaying = false;
    _delayService.clearCycles();
  }

  bool isSameTime(double time1, double time2) {
    const precision = 0.001;
    return (time1 - time2).abs() < precision;
  }

  void changeSpeed(Speed newSpeed) {
    _delayService.changeSpeed(newSpeed);
    _soundService.setSpeed(newSpeed);
  }

  void tick() {
    //print("actionIndex: $actionIndex");
    //print("replayActions.length: ${replayActions.length}");
    if (_gameService.recordedTimeIndex >= _gameService.recordedTimes.length) {
      pause();
      return;
    }
    _gameService.nextRecordedTime();
    double currentTime =
        _gameService.recordedTimes[_gameService.recordedTimeIndex];
    _replayTimeController.add(currentTime);
    do {
      ReplayAction replayAction = replayActions[actionIndex];
      if (isSameTime(replayAction.time,
          _gameService.recordedTimes[_gameService.recordedTimeIndex])) {
        //print("replayAction.time: ${replayAction.time}");
        actionIndex++;
        doAction(replayAction.category, replayAction.input);
      }
    } while (actionIndex < replayActions.length &&
        isSameTime(replayActions[actionIndex].time,
            _gameService.recordedTimes[_gameService.recordedTimeIndex]));
  }

  void doReplay() {
    _delayService.timeIsPaused = false;
    _gameService.isReplayMode = true;
    _gameService.recordedTimeIndex = 0;
    // replayEvent.emit(); // This is meant to trigger the replay in other components
    isReplayMode = true;
    isReplaying = true;
    storeActions = false;
    actionIndex = 0;
    _delayService.doCyclically(NB_MS_IN_SECOND * TICK, () {
      //print("replayActions.length: ${replayActions.length}");
      if (actionIndex < replayActions.length &&
          _gameService.recordedTimeIndex < _gameService.recordedTimes.length) {
        tick();
        print("tickcccccccccccccccccccccccccccccccccccccccc");
        print(DelayService.instance.speed);
      } else {
        pause();
      }
    });
  }

  @override
  void dispose() {
    closeStreams();
    // replayEvent.close();
    // replayActionTrigger.close();
    // _replayTimeController.close();
    super.dispose();
  }
}
